tools/castle-editor/code/castleundosystem.pas

```diff
@@ -0,0 +1,276 @@
+  end;
+
+begin
+  Result := Self.InstanceSize + SizeOfAnsiString(Data) + SizeOfAnsiString(Selected) + SizeOfAnsiString(Comment);
+end;
+
+constructor TUndoSystem.Create(AOwner: TComponent);
+begin
+  inherited;
+  UndoHistory := TUndoHistory.Create(true);
+  ClearUndoHistory;
+end;
+
+destructor TUndoSystem.Destroy;
+begin
+  FreeAndNil(UndoHistory);
+  inherited;
+end;
+
+function TUndoSystem.UndoHistorySize: Integer;
+var
+  U: TUndoHistoryElement;
+begin
+  Result := 0;
+  for U in UndoHistory do
+    Result += U.Size;
+end;
+
+procedure TUndoSystem.RecordUndo(const UndoData: TUndoData; const SelectedComponent: TSelectedComponent; const ItemIndex: Integer; const TabIndex: Integer; const UndoComment: String);
+var
+  NewUndoElement: TUndoHistoryElement;
+  I: Integer;
+  NewUndoHistorySize: Integer;
+begin
+  ScheduleRecordUndoOnRelease := false;
+  if (UndoHistory.Count > 0) and (UndoData = UndoHistory[CurrentUndo].Data) then
+  begin
+    if (SelectedComponent = UndoHistory[CurrentUndo].Selected) or (UndoHistory[CurrentUndo].Selected = '') then
+      WriteLnLog('New Undo is identical to previous Undo record. Not saving.')
+    else
+    begin
+      UndoHistory[CurrentUndo].Selected := SelectedComponent;
+      WriteLnLog('New Undo is identical to previous Undo record. Only selection has changed from ' + UndoHistory[CurrentUndo].Selected + ' to ' + SelectedComponent + '. Not saving.');
+      OnUpdateUndo(Self);
+    end;
+    if (UndoComment <> '') and (UndoHistory[CurrentUndo].Comment = '') then
+    begin
+      WriteLnLog('The last undo comment was set to generic "Edit value", overwriting with better reason: ' + UndoComment);
+      UndoHistory[CurrentUndo].Comment := UndoComment;
+      OnUpdateUndo(Self);
+    end;
+    Exit;
+  end;
+  WriteLnLog('Saving Undo record. CurrentUndo = ' + IntToStr(CurrentUndo));
+  //Clean all next undo records if available;
+  for I := UndoHistory.Count - 1 downto CurrentUndo + 1 do
+    UndoHistory.Delete(I);
+  //add new UndoElement
+  NewUndoElement := TUndoHistoryElement.Create;
+  NewUndoElement.Data := UndoData;
+  NewUndoElement.Selected := SelectedComponent;
+  NewUndoElement.ItemIndex := ItemIndex;
+  NewUndoElement.TabIndex := TabIndex;
+  NewUndoElement.Comment := UndoComment;
+  UndoHistory.Add(NewUndoElement);
+  Inc(CurrentUndo);
+  Assert(UndoHistory.Count - 1 = CurrentUndo);
+  //make sure Undo doesn't consume too much RAM
+  {if CurrentUndo > MaxUndo then
+  begin
+    UndoHistory.Delete(0);
+    Dec(CurrentUndo);
+  end;}
+  NewUndoHistorySize := UndoHistorySize;
+  while (NewUndoHistorySize > MaxUndoHistorySize) and (UndoHistory.Count > 1) do
+  begin
+    NewUndoHistorySize -= UndoHistory[0].Size;
+    UndoHistory.Delete(0);
+    Dec(CurrentUndo);
+  end;
+  Assert(UndoHistorySize = NewUndoHistorySize);
+  OnUpdateUndo(Self);
+  WriteLnLog('Undo record saved. CurrentUndo = ' + IntToStr(CurrentUndo) + '; Undo History Size = ' + IntToStr(NewUndoHistorySize div 1024) + 'kb.');
```

Comment:
Instead of 

```
IntToStr(NewUndoHistorySize div 1024) + 'kb'
```

use 

```
SizeToStr(NewUndoHistorySize)
```

from CastleStringUtils. It automatically chooses correct units (like KB or MB).

Reply:
Done!